File: packages/mcp-server/src/tools/epics.ts
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { pmApi, PmApiError } from '../pm-api.js';

function err(msg: string) { return { content: [{ type: 'text' as const, text: `Error: ${msg}` }] }; }
function ok(data: unknown) { return { content: [{ type: 'text' as const, text: JSON.stringify(data, null, 2) }] }; }
function handleError(e: unknown) { return err(e instanceof PmApiError ? e.message : String(e)); }

export function registerEpicsTools(server: McpServer) {
  server.tool('list_epics', 'List epics', {
    project_id: z.string().describe('Project ID'),
  }, async ({ project_id }) => {
    try {
      const data = await pmApi(`/api/epics?project_id=${encodeURIComponent(project_id)}`);
      return ok(data);
    } catch (e) { return handleError(e); }
  });

  server.tool('add_epic', 'Create epic', {
    project_id: z.string(),
    title: z.string(),
    priority: z.enum(['low', 'medium', 'high', 'critical']).optional(),
    description: z.string().optional(),
  }, async (body) => {
    try {
      const data = await pmApi('/api/epics', { method: 'POST', body: JSON.stringify(body) });
      return ok(data);
    } catch (e) { return handleError(e); }
  });

  server.tool('update_epic', 'Update epic', {
    epic_id: z.string(),
    title: z.string().optional(),
    status: z.string().optional(),
    priority: z.string().optional(),
    description: z.string().optional(),
  }, async ({ epic_id, ...updates }) => {
    try {
      const data = await pmApi(`/api/epics/${encodeURIComponent(epic_id)}`, {
        method: 'PATCH',
        body: JSON.stringify(updates),
      });
      return ok(data);
    } catch (e) { return handleError(e); }
  });

  server.tool('delete_epic', 'Delete epic', {
    epic_id: z.string(),
  }, async ({ epic_id }) => {
    try {
      await pmApi(`/api/epics/${encodeURIComponent(epic_id)}`, { method: 'DELETE' });
      return ok({ deleted: true });
    } catch (e) { return handleError(e); }
  });
}

File: backend/sprintable_mcp/tools/epics.py
"""에픽 관련 MCP 도구 (3개) — E-SECURITY SEC-S1 확장: delete_epic 제거(에이전트 hard-delete 차단,
delete_story와 동형 조치. 까심 적대적 QA 발견 갭 — cascade로 소속 stories까지 물리삭제)."""
from __future__ import annotations

from mcp.types import TextContent

from ..api_client import client
from ..response import err, ok
from ..schemas import EpicStatus, SprintableInput, StoryPriority


class ListEpicsInput(SprintableInput):
    pass


class AddEpicInput(SprintableInput):
    title: str
    priority: StoryPriority | None = None
    description: str | None = None
    objective: str | None = None
    success_criteria: str | None = None
    target_sp: int | None = None
    target_date: str | None = None


class UpdateEpicInput(SprintableInput):
    epic_id: str
    title: str | None = None
    status: EpicStatus | None = None
    priority: StoryPriority | None = None
    description: str | None = None
    objective: str | None = None
    success_criteria: str | None = None
    target_sp: int | None = None
    target_date: str | None = None


async def list_epics(args: ListEpicsInput) -> list[TextContent]:
    """에픽 목록 조회."""
    try:
        return ok(await client.get("/api/v2/epics", params={"project_id": client.project_id}))
    except Exception as exc:
        return err(str(exc))


async def add_epic(args: AddEpicInput) -> list[TextContent]:
    """에픽 생성."""
    body: dict = {"title": args.title, "project_id": client.project_id}
    if args.priority:
        body["priority"] = args.priority.value
    for field in ("description", "objective", "success_criteria", "target_date"):
        val = getattr(args, field)
        if val is not None:
            body[field] = val
    if args.target_sp is not None:
        body["target_sp"] = args.target_sp
    try:
        return ok(await client.post("/api/v2/epics", json=body))
    except Exception as exc:
        return err(str(exc))


async def update_epic(args: UpdateEpicInput) -> list[TextContent]:
    """에픽 수정."""
    updates: dict = {}
    if args.title is not None:
        updates["title"] = args.title
    if args.status is not None:
        updates["status"] = args.status.value
    if args.priority is not None:
        updates["priority"] = args.priority.value
    for field in ("description", "objective", "success_criteria", "target_date"):
        val = getattr(args, field)
        if val is not None:
            updates[field] = val
    if args.target_sp is not None:
        updates["target_sp"] = args.target_sp
    try:
        return ok(await client.patch(f"/api/v2/epics/{args.epic_id}", json=updates))
    except Exception as exc:
        return err(str(exc))
